src/Simplex/Messaging/Agent/Store.hs

```diff
@@ -84,33 +122,100 @@ instance Eq SomeConn where
 
 deriving instance Show SomeConn
 
-data MessageDelivery = MessageDelivery
+-- * Message types
+
+-- | A message in either direction that is stored by the agent.
+data Msg = MRcv RcvMsg | MSnd SndMsg
+  deriving (Eq, Show)
+
+-- | A message received by the agent from a sender.
+data RcvMsg = RcvMsg
+  { baseData :: MsgBase,
+    internalRcvId :: InternalRcvId,
+    -- | Id of the message at sender, corresponds to `internalSndId` from the sender's side.
+    externalSndId :: ExternalSndId,
+    externalSndTs :: ExternalSndTs,
+    brokerId :: BrokerId,
+    brokerTs :: BrokerTs,
+    rcvStatus :: RcvStatus,
+    -- | Timestamp of acknowledgement to broker, corresponds to `AcknowledgedToBroker` status.
+    -- Do not mix up with `brokerTs` - timestamp created at broker after it receives the message from sender.
+    ackBrokerTs :: AckBrokerTs,
+    -- | Timestamp of acknowledgement to sender, corresponds to `AcknowledgedToSender` status.
+    -- Do not mix up with `externalSndTs` - timestamp created at sender before sending,
+    -- which in its turn corresponds to `internalTs`.
+    ackSenderTs :: AckSenderTs
+  }
+  deriving (Eq, Show)
+
+type InternalRcvId = Int64
+
+type ExternalSndId = Integer
+
+type ExternalSndTs = UTCTime
+
+type BrokerId = MsgId
+
+type BrokerTs = UTCTime
+
+data RcvStatus
+  = Received
+  | AcknowledgedToBroker
+  | AcknowledgedToSender
+  deriving (Eq, Show)
+
+type AckBrokerTs = UTCTime
+
+type AckSenderTs = UTCTime
+
+-- | A message sent by the agent to a recipient.
+data SndMsg = SndMsg
+  { baseData :: MsgBase,
+    -- | Id of the message sent / to be sent, as in its number in order of sending.
+    internalSndId :: InternalSndId,
+    sndStatus :: SndStatus,
+    -- | Timestamp of the message received by broker, corresponds to `Sent` status.
+    sentTs :: SentTs,
+    -- | Timestamp of the message received by recipient, corresponds to `Delivered` status.
+    deliveredTs :: DeliveredTs
+  }
+  deriving (Eq, Show)
+
+type InternalSndId = Int64
+
+data SndStatus
+  = Created
+  | Sent
+  | Delivered
+  deriving (Eq, Show)
+
+type SentTs = UTCTime
+
+type DeliveredTs = UTCTime
+
+-- | Base message data independent of direction.
+data MsgBase = MsgBase
   { connAlias :: ConnAlias,
-    agentMsgId :: Int,
-    timestamp :: UTCTime,
-    message :: AMessage,
-    direction :: QueueDirection,
-    msgStatus :: DeliveryStatus
+    -- | Monotonically increasing id of a message per connection, internal to the agent.
+    -- Preserves ordering between both received and sent messages.
+    internalId :: InternalId,
+    internalTs :: InternalTs,
+    msgBody :: MsgBody
```

Comment:
no need to change in this PR, but it feels like this structure misses the direction tag to be "a message". It is kind of stupid to add it of course, given that we have this tag on the level above as constructor. Let's merge and see how it is used and change later. Names can the be revisited later too.